NVIDIA/deepops
# Introduce nfs-client-provisioner to K8s deployment and include default NFS server++

We have had several changes with rook-ceph lately and that is several versions behind. We also have had many requests for easier NFS integration.

The goal of this PR is to alleviate these issues by deploying all Kubernetes cluster with a Default StorageClass that can be used to deploy applications such as Kubeflow, store/share data for multi-node jobs, and provide NFS access.

With this PR we are not removing rook-ceph, but are moving more workloads/workflows over to use NFS instead of the CephBlock StorageClass. This will have the added benefit of supporting MPIJobs in Kubeflow.

With this PR we will also be updating Prometheus to persist data by default.

With this PR there are also a few fixes and major enhancements to Kubeflow

**Major changes to K8s deployment:**
* `mgmt01` will be setup as an NFS server by default when running `k8s-cluster.yml`
* `nfs-client-provisioner` will use the `mgmt01` NFS server as a default storage class
* `scripts/k8s/deploy_monitoring.sh` will require a default storageclass and persist data (with option to override)
* Kubeflow now launches with NGC containers as the default
* Kubeflow now launches with custom configurations saved in `workloads/service/k8s/kubeflow-install`

**New configuration options:**
_group_vars:_
k8s_nfs_server: The address of the existing NFS server or mgmt01
k8s_nfs_export_path: The existing export path to use in nfs-client and/or new export path to set up.
nfs_exports: Now set by default in k8s-cluster groupvars.

_skip_tags:_
nfs_server:  Skip setting up an NFS server
nfs_client: Skip mount NFS on clients
nfs_software: Skip installing nfsutils on all hosts
nfs_mkdir: Skip `mkdir -p $(export_path) && chmod 777 $(export_path)`
nfs_client_provisioner: Skip helm install

-scripts:_
`deploy_monitoring ` `-x`: Deploy monitoring stack using legacy config with no persistent storage.

**Implementation details:**
- [x] Create a role to perform helm install of nfs-client-provisioner
- [x] Update k8s-cluster documentation & update Kubeflow documentation
- [x] Update Kubeflow script (documentation and bugfixes)
- [x] Add non-default installation mode to rook-ceph
- [x] Add option to `skip-tags` nfs_client, nfs_server, and nfs_client_provisioner
- [x] Update prometheus to use a PV for data storage (and add additional tests)
- [x ] ~~Move helm install off the `stable/nfs-client-provisioner` repo~~ Implement via local files and git submodule
- [x] Update Jenkins testing 


**Test Plan:**
1. Verify K8s install default works and supports Kubeflow

In a cluster with at least 1 mgmt node and 1 worker node Install K8s using the standard `k8s-cluster.yml`. After installing (and without install rook) install Kubeflow using the `deploy_kubeflow.sh`. Verify that kubeflow UI comes up, the mysql pod is running, and `ls /export/shared` to verify storage was created in NFS.

2. Verify K8s install with existing NFS is supported

In a cluster with a NFS server, at least 1 mgmt node and 1 worker node update `config/group_vars/k8s-cluster.yml` update `k8s_nfs_server` and `k8s_nfs_export_path` with the information from the nfs server. Install K8s using the standard `k8s-cluster.yml` and specify `--skip-tags=nfs_server`. After installing (and without install rook) install Kubeflow using the `deploy_kubeflow.sh`. Verify that kubeflow UI comes up, the mysql pod is running, and `ls /export/shared` to verify storage was created in NFS.


3. Verify K8s install with no NFS is supported

In a cluster with at least 1 mgmt node and 1 worker node Install K8s using the standard `k8s-cluster.yml` and specify `--skip-tags=k8s_nfs_client_provisioner`. After installing deploy rook with `deploy_rook.sh && deploy_rook.sh -w` and install Kubeflow using the `deploy_kubeflow.sh`. Verify that kubeflow UI comes up, the mysql pod is running, and `ls /export/shared` to verify storage was created in NFS.

4. Verify K8s install with NFS & Rook

In a cluster with at least 1 mgmt node and 1 worker node Install K8s using the standard `k8s-cluster.yml`. After installing deploy rook with `deploy_rook.sh && deploy_rook.sh -w` install Kubeflow using the `deploy_kubeflow.sh`. Verify that kubeflow UI comes up, the mysql pod is running, and `ls /export/shared` to verify storage was created in NFS. Verify both storage classes with `kubectl get storageclass`

5. Verify all Kubeflow containers work

After Kubeflow is up and running, log in and launch one of each container. Verify they come up and the tutorial notebooksa are displayed.
